src/utils/flags.cpp

```diff
@@ -73,6 +78,11 @@ class flag_data
             value<type>() = old_val;                                                               \
             return error_s::make(ERR_INVALID_PARAMETERS, "value validation failed");               \
         }                                                                                          \
+        std::string total_message;                                                                 \
```

Comment:
```
        std::string total_message;                                                                 \
        if (_group_validators_runner && !_group_validators_runner(total_message)) {                \
            value<type>() = old_val;                                                               \
            return error_s::make(ERR_INVALID_PARAMETERS, total_message.c_str());                   \
        }                  
```
I think it's better to move group validator to `flag_registry::update_flag`

Reply:
Initially I put the detection of group validators in `flag_registry::update_flag`. However, the old value has to be read and saved in a temporary variable, since later once the detection failed, the old value will be used to set back to  `flag_data` again. We will read the old value by template, which has been implemented well in `FLAG_DATA_UPDATE_CASE`. So can we reuse the template in `FLAG_DATA_UPDATE_CASE` ?

Reply:
OK.

Reply:
But I think you can pass the pointer of flag_registry to flag_data, so you can use flag_registry::run_group_validator here

Reply:
If we want to use the pointer of `flag_registry` and flag_registry::run_group_validators in `flag_data`, we have to declare class `flag_registry` before `flag_data`.  Actually flag_registry also depends on `flag_data`. Thus, the declaration of `flag_data` and `flag_registry` should be moved to the header file. This will lead to a big change to current structure.

Reply:
I think it's OK to change the current structure. What we should do is split `flag_data` and `flag_registry` into seperate files.
The reason why I want to do this is that I think group flag validators belongs to `flag_registry`. It's better to not 
 to affect `flag_data`

Reply:
Yeah I agree that it's better not to affect `flag_data`. All flags and grouped validators are saved in `flag_registry`. So at first I put everything about grouped validators in `flag_registry`. `flag_data` knows nothing about grouped validators. Even it knows nothing about `flag_registry`.

However,  the value that grouped validators detect belong to each `flag_data`, just as the individual validator. In this respect, there must be some "relation" between `flag_data` and grouped validators. I try to think other alternative patterns, but only `flag_data` knows how to parse to the templated value. Also the error message is defined in the returned `error_s`. So I think running grouped validators should be done by `update` function.

I think an extra advantage of taking `flag_registry` pointer in `flag_data` is that if later there's another "relation" found, `flag_registry` pointer can be reused.

Reply:
Suggested by @levy5307 , we will use `join_point` to trigger group validations while `flag_data` is updating the value.